Not possible start Spring Cloud Data Flow in Windows

Copied from https://github.com/spring-cloud/spring-cloud-dataflow/issues/1580

In `Windows 10 Home` edition, I have configured the `JAVA_HOME` and `PATH` variables (same approach used since `Windows XP`) all my app that use Java work fine, same appreciation for any tool that needs Java, such as (IDE, LibreOffice, etc...) 

Observe the following: 

    > javac -version
    javac 1.8.0_131
    
    > java -version
    java version "1.8.0_131"
    Java(TM) SE Runtime Environment (build 1.8.0_131-b11)
    Java HotSpot(TM) 64-Bit Server VM (build 25.131-b11, mixed mode)

According with [Spring Cloud Data Flow][1]

 When I do `java -jar spring-cloud-dataflow-server-local-1.2.2.RELEASE.jar` the following happens:

    > java -jar spring-cloud-dataflow-server-local-1.2.2.RELEASE.jar
      ____                              ____ _                __
     / ___| _ __  _ __(_)_ __   __ _   / ___| | ___  _   _  __| |
     \___ \| '_ \| '__| | '_ \ / _` | | |   | |/ _ \| | | |/ _` |
      ___) | |_) | |  | | | | | (_| | | |___| | (_) | |_| | (_| |
     |____/| .__/|_|  |_|_| |_|\__, |  \____|_|\___/ \__,_|\__,_|
      ____ |_|    _          __|___/                 __________
     |  _ \  __ _| |_ __ _  |  ___| | _____      __  \ \ \ \ \ \
     | | | |/ _` | __/ _` | | |_  | |/ _ \ \ /\ / /   \ \ \ \ \ \
     | |_| | (_| | || (_| | |  _| | | (_) \ V  V /    / / / / / /
     |____/ \__,_|\__\__,_| |_|   |_|\___/ \_/\_/    /_/_/_/_/_/
    
    Spring Cloud Data Flow Local Server  (v1.2.2.RELEASE)
    
    2017-07-01 14:43:37.512  INFO 2260 --- [           main] c.c.c.ConfigServicePropertySourceLocator : Fetching config from server at: http://localhost:8888
    2017-07-01 14:43:38.839  WARN 2260 --- [           main] c.c.c.ConfigServicePropertySourceLocator : Could not locate PropertySource: I/O error on GET request for "http://localhost:8888/spring-cloud-dataflow-server-local/default": Connection refused: connect; nested exception is java.net.ConnectException: Connection refused: connect
    2017-07-01 14:43:38.843  INFO 2260 --- [           main] o.s.c.d.s.local.LocalDataFlowServer      : No active profile set, falling back to default profiles: default
    2017-07-01 14:43:42.687  INFO 2260 --- [           main] .s.d.r.c.RepositoryConfigurationDelegate : Multiple Spring Data modules found, entering strict repository configuration mode!
    2017-07-01 14:43:43.614  INFO 2260 --- [           main] o.s.cloud.context.scope.GenericScope     : BeanFactory id=b15ff760-91d3-31de-a2a8-f2cc5ce41a5c
    2017-07-01 14:43:46.475  INFO 2260 --- [           main] o.apache.catalina.core.StandardService   : Starting service [Tomcat]
    2017-07-01 14:43:46.478  INFO 2260 --- [           main] org.apache.catalina.core.StandardEngine  : Starting Servlet Engine: Apache Tomcat/8.5.15
    2017-07-01 14:43:46.730  INFO 2260 --- [ost-startStop-1] o.a.c.c.C.[Tomcat].[localhost].[/]       : Initializing Spring embedded WebApplicationContext
    2017-07-01 14:43:47.871  INFO 2260 --- [           main] erverConfiguration$H2ServerConfiguration : Starting H2 Server with URL: jdbc:h2:tcp://localhost:19092/mem:dataflow
    2017-07-01 14:43:49.381  INFO 2260 --- [           main] o.s.c.d.s.r.s.DataflowRdbmsInitializer   : Adding dataflow schema classpath:schema-h2-common.sql for h2 database
    2017-07-01 14:43:49.382  INFO 2260 --- [           main] o.s.c.d.s.r.s.DataflowRdbmsInitializer   : Adding dataflow schema classpath:schema-h2-streams.sql for h2 database
    2017-07-01 14:43:49.382  INFO 2260 --- [           main] o.s.c.d.s.r.s.DataflowRdbmsInitializer   : Adding dataflow schema classpath:schema-h2-tasks.sql for h2 database
    2017-07-01 14:43:49.383  INFO 2260 --- [           main] o.s.c.d.s.r.s.DataflowRdbmsInitializer   : Adding dataflow schema classpath:schema-h2-deployment.sql for h2 database
    2017-07-01 14:43:49.393  INFO 2260 --- [           main] o.s.jdbc.datasource.init.ScriptUtils     : Executing SQL script from class path resource [schema-h2-common.sql]
    2017-07-01 14:43:49.420  INFO 2260 --- [           main] o.s.jdbc.datasource.init.ScriptUtils     : Executed SQL script from class path resource [schema-h2-common.sql] in 27 ms.
    2017-07-01 14:43:49.420  INFO 2260 --- [           main] o.s.jdbc.datasource.init.ScriptUtils     : Executing SQL script from class path resource [schema-h2-streams.sql]
    2017-07-01 14:43:49.427  INFO 2260 --- [           main] o.s.jdbc.datasource.init.ScriptUtils     : Executed SQL script from class path resource [schema-h2-streams.sql] in 4 ms.
    2017-07-01 14:43:49.427  INFO 2260 --- [           main] o.s.jdbc.datasource.init.ScriptUtils     : Executing SQL script from class path resource [schema-h2-tasks.sql]
    2017-07-01 14:43:49.434  INFO 2260 --- [           main] o.s.jdbc.datasource.init.ScriptUtils     : Executed SQL script from class path resource [schema-h2-tasks.sql] in 7 ms.
    2017-07-01 14:43:49.435  INFO 2260 --- [           main] o.s.jdbc.datasource.init.ScriptUtils     : Executing SQL script from class path resource [schema-h2-deployment.sql]
    2017-07-01 14:43:49.438  INFO 2260 --- [           main] o.s.jdbc.datasource.init.ScriptUtils     : Executed SQL script from class path resource [schema-h2-deployment.sql] in 2 ms.
    2017-07-01 14:43:50.946  INFO 2260 --- [           main] o.s.b.c.r.s.JobRepositoryFactoryBean     : No database type set, using meta data indicating: H2
    2017-07-01 14:43:51.414  INFO 2260 --- [           main] o.s.jdbc.datasource.init.ScriptUtils     : Executing SQL script from class path resource [org/springframework/batch/core/schema-h2.sql]
    2017-07-01 14:43:51.492  INFO 2260 --- [           main] o.s.jdbc.datasource.init.ScriptUtils     : Executed SQL script from class path resource [org/springframework/batch/core/schema-h2.sql] in 78 ms. 2017-07-01 14:43:51.521  INFO 2260 --- [           main] o.s.jdbc.datasource.init.ScriptUtils     : Executing SQL script from class path resource [org/springframework/cloud/task/schema-h2.sql]
    2017-07-01 14:43:51.538  INFO 2260 --- [           main] o.s.jdbc.datasource.init.ScriptUtils     : Executed SQL script from class path resource [org/springframework/cloud/task/schema-h2.sql] in 16 ms.

...

    Caused by: org.springframework.beans.factory.BeanCreationException: 
      Error creating bean with name 'spring.cloud.deployer.local-org.springframework.cloud.deployer.spi.local.LocalDeployerProperties': 
      Instantiation of bean failed; nested exception is org.springframework.beans.BeanInstantiationException: 
      Failed to instantiate [org.springframework.cloud.deployer.spi.local.LocalDeployerProperties]: 
      Constructor threw exception; nested exception is java.lang.IllegalArgumentException: 
      Java executable discovered via 'java.home' system property 'C:\whiteroom\java\jre1.8.0_131' is not executable or does not exist.
      ...
      Caused by: java.lang.IllegalArgumentException: Java executable discovered via 'java.home' system property 'C:\whiteroom\java\jre1.8.0_131' is not executable or does not exist.
            at

--- a/spring-cloud-deployer-local/src/main/java/org/springframework/cloud/deployer/spi/local/LocalDeployerProperties.java
+++ b/spring-cloud-deployer-local/src/main/java/org/springframework/cloud/deployer/spi/local/LocalDeployerProperties.java
@@ -51,7 +51,7 @@ public class LocalDeployerProperties {
 
 	private static final Logger logger = LoggerFactory.getLogger(LocalDeployerProperties.class);
 
-	private static final String JAVA_COMMAND = "java";
+	private static final String JAVA_COMMAND = isWindows() ? "java.exe" : "java";
 
 	/**
 	 * Directory in which all created processes will run and create log files.
@@ -139,16 +139,32 @@ public class LocalDeployerProperties {
 		String javaExecutablePath = JAVA_COMMAND;
 		String javaHome = System.getProperty("java.home");
 		if (javaHome != null) {
-			File javaExecutable = new File(javaHome, "bin/" + javaExecutablePath);
-			Assert.isTrue(javaExecutable.canExecute(), "Java executable discovered via 'java.home' system property '"
-					+ javaHome + "' is not executable or does not exist.");
+			File javaExecutable = new File(javaHome, "bin" + File.separator + javaExecutablePath);
+			Assert.isTrue(javaExecutable.canExecute(),
+					"Java executable'" + javaExecutable + "'discovered via 'java.home' system property '" + javaHome
+							+ "' is not executable or does not exist.");
 			javaExecutablePath = javaExecutable.getAbsolutePath();
 		}
 		else {
-			logger.warn("System property 'java.home' is not set. Defaulting to the java executable path as 'java' assuming it's in PATH.");
+			logger.warn("System property 'java.home' is not set. Defaulting to the java executable path as "
+					+ JAVA_COMMAND + " assuming it's in PATH.");
 		}
 
 		return javaExecutablePath;
+	}
+
+	private static boolean isWindows() {
+		String osName = null;
+		try {
+			osName = System.getProperty("os.name");
+		} catch (Exception e) {
+		}
+		if (osName == null) {
+			return false;
+		}
+		else {
+			return osName.toLowerCase().startsWith("windows");
+		}
 	}
 
 	@Override

